### ISSUE TITLE
how to simply compile like before?

### ISSUE THREAD

**Issue by ROBERT-MCDOWELL** [gsliepen/tinc#374]
how to compile with the previous configure options like
--enable-jumbograms --enable-tunemu --enable-uml
?
thanks

**Comment by gsliepen** [comment]
You now have to specify it when running meson like so:
```
meson build -Djumbograms=true -Dtunemu=enabled -Duml=true
```
If you already already ran `meson build` before, just add the `--reconfigure` option to reconfigure that build directory with the desired options. To get a list of the supported options, run `meson configure`.

**Comment by gsliepen** [comment]
I hope this answers the question. But keeping this open as this definitely should be added to the documentation.

**Comment by ROBERT-MCDOWELL** [comment]
thank you very much gsliepen

**Comment by ROBERT-MCDOWELL** [comment]
I did meson build -Djumbograms=true -Dtunemu=enabled -Duml=true after deleted the build folder and compiled well.
then should I go to build folder and ???

**Comment by hg** [comment]
It's already described [here][install], but maybe it can be stripped down to absolute bare bones and moved further up.

[install]: https://github.com/gsliepen/tinc/blob/1.1/INSTALL.md#native

---

@ROBERT-MCDOWELL it's not very clear what issues you're having. With meson, you first create a build directory, passing any configuration flags in the process:

    $ meson foobar -D jumbograms=true

And then compile it with

    $ ninja -C foobar

where `foobar` is the name of your build directory. It can be named however you like, I'm using `foobar` here to avoid confusion.

If everything goes well, you'll find binaries inside `foobar`:

    $ ./foobar/src/tinc -c /tmp/foo init
    $ ./foobar/src/tincd -c /tmp/foo -D

You can either copy them from there manually, or run

    $ meson install -C foobar

although it's not the best idea (better to use your distribution's package manager).

---

To see the full list of available options, run:

    $ meson configure

**Comment by ROBERT-MCDOWELL** [comment]
ha ok, I was missing meson install -C foobar
to install it in /usr/local/bin.... I prefer to use my own compiled tinc than the distribution one btw...
thanks!

**Comment by SilverBut** [comment]
Jesus... why use meson (or any other tool)? I just witnessed this project's building infra inflated from simple Makefile to Ninja to Meson, while the function list and bug list has not grown so much.

I can see the importance of having a good infra, as this can make PR being reviewed easier (with more unit tests), and allowing more rapid version changes. But I'm not sure changing the system from X to Y is the best method.

But, well, since it has been merged, I think I would better install meson on my server.

**Comment by fangfufu** [comment]
I am a bit concerned that tinc might have problem with building on router-oriented OS, such as OpenWRT. I guess we will just have to wait and see.

**Comment by eli-schwartz** [comment]
> I just witnessed this project's building infra inflated from simple Makefile to Ninja to Meson, while the function list and bug list has not grown so much.

This project has never used a simple Makefile though. The first commit included autotools configure.in, which is a full-fledged build system that generates Makefiles.

So it inflated from "autotools and make" to "ninja and meson" (and that second stage is one combined, no one used bare ninja anymore than they used bare Make here).

Autotools has its own complexity, and the strongest argument people bring for its simplicity is that it's "simple for the end user to use a prepared tarball". Of course it's only simple on non-Windows platforms...

> I am a bit concerned that tinc might have problem with building on router-oriented OS, such as OpenWRT. I guess we will just have to wait and see.

Since multiple people are expressing concern about this, I figured I'd come out and ask: what exactly is the worry about using meson? Is it not portable to your OS? (It uses python, more or less, python is pretty portable.) Is it too big for space-constrained systems? Do you need to cross-compile?

What's the use case you have? That's the most important piece of information for determining whether it is feasible/practical to build with meson.

**Comment by hg** [comment]
@SilverBut

meson is used by so many projects these days that it's going to be present on any build server, and it's pretty lightweight in any case (see below). If we consider UNIX-oriented 'system' software, it's probably the most popular build system for anything written in the last 10-15 years. Any package maintainer knows how to use it.

A few things it gives us in return from the top of my head:

- it's easy to work with: adding support for a new library or feature produces a small, clean, easily readable diff (latest [example][sd]); with more complex logic it's especially noticeable. No more 6-argument m4 macros where you need to search the manual for whether to put the name of the function into the third, or into the fifth argument, no more dragging around 12 copies of 200-line m4 library-specific macros. If you don't want this project to die in the long term, *do care* about not putting more roadblocks in the way of future contributors.
- has equally good windows support (many do care about that one)
- precily because autotools is not "a makefile", building tinc doesn't depend on UNIX-y environment and shell tools anymore, so MSVC now works (the compiler of choice for most Windows developers; other weird compilers could be added if anyone requests it)
- [dependency wrapping][wrap]: much easier to compile for Windows or other awkward platforms where it is difficult/impossible to install pre-built libraries. What previously required cloning, configuring, and building five separate projects (`lz4` + `lzo2` + `openssl` + `zlib` + `tinc`) now can be done in a single command.
- does things you'd expect to work out of the box these days: generates proper compilation database without unreliable third-party hacks (used by `clangd` and `clang-tidy`); uses ccache if it's present; runs tests in parallel with actually working timeouts, and so on.
- proper support for out-of-tree builds (I use it all the time for testing different configurations/building for different targets, it's also heavily used by CI)

[sd]: https://github.com/gsliepen/tinc/commit/d2949381713c35210a43391524494c639f6f1d48#diff-d6a58e7a5653d78e8d2721c2f29d500b1099b86f742dca74306d076a32a6a89e
[wrap]: https://mesonbuild.com/Wrap-dependency-system-manual.html

autotools are definitely not "a simple makefile". The build files it generates are exceedingly complex: dozens of files that call each other, tens of thousands of lines of shell code and m4 macros that are impossible to debug when anything goes wrong; it's the opposite of simple.

You needed a bunch of separate tools to build the project (`make` + `autoconf` + `automake`;  I recollect it having issues with BSD make on some BSDs, so you might actually need `gmake` there). Now you need `meson` + `ninja` + preferably `pkgconf`. What's the difference?

You also needed to remember of a bunch of weird incantations, I'd argue worse than it is now (wtf is `autoreconf -fsi && ./configure --foo --bar && make -j$(nproc) && make install`? is it better or worse than `meson build && meson install -C build`?). By the way, `nproc` is not available everywhere, so that command would look a bit different on other platforms.

Let's see how much space both require in a freshly created Debian container (`debian:bullseye-slim`):

```
# apt install automake autoconf make
...
Need to get 9190 kB of archives.
After this operation, 53.3 MB of additional disk space will be used.

# apt install meson ninja-build
...
Need to get 8176 kB of archives.
After this operation, 28.7 MB of additional disk space will be used.
```

And that's an empty system without Python 3, which is going to be pre-installed on any desktop or server system already. Let's install it and see how heavy `meson` + `ninja` are:

```
# apt install python3
...
Need to get 7437 kB of archives.
After this operation, 24.9 MB of additional disk space will be used.
...
done.

# apt install meson ninja-build
...
Need to get 738 kB of archives.
After this operation, 3764 kB of additional disk space will be used.
```

OK, you'd also want `pkgconf`, but the package is something like 70 KBs (30 + 40 to be more precise). It pulls in a compiler (which you'd need with any build system), so I'm not testing it here.

autotools is pretty buggy even after 30 years of development precisely because it's not a "simple makefile". I can't count the number of clean & reconfigure I had to do in day-to-day development because it blew it brains out and either stopped noticing changes in a source file, or stopped working altogether (and that's Linux, the best and most stable platform for such things).

---

@fangfufu we've been reading lots of concerns (let's call it how it is: FUD), but not many actual issues yet. IIUC @lancethepants already successfully built the latest version for his tomato project. There were a [couple of small issues][lp] (because it wasn't much tested on routers), which were quickly fixed.

I'll happily add more cross-compilation jobs to CI so building for routers is always tested, if you'd point me to a working toolchain, and (preferably) a very short instruction on how to do it if there are any particulars to be aware of. I couldn't find anything like that on the website, and I don't have time or interest to dig deeply into that subject tbh.

[lp]: https://github.com/gsliepen/tinc/issues?q=is%3Aissue+author%3Alancethepants

**Comment by gsliepen** [comment]
I'll close this ticket again, Robert's original question has been answered, and since then the documentation has improved. @SilverBut if there are any problems you are having while trying to compile tinc with the new build system, don't hesitate to create a new issue here!

**Comment by gsliepen** [comment]
@SilverBut What OS are those 3 special nodes running? Because creating a ticket for supporting a particular OS or CPU architecture would perhaps be appropriate. Also consider that it might actually be easier to [cross-compile](https://mesonbuild.com/Cross-compilation.html) tinc for those special nodes with meson than with autoconf.

**Comment by SilverBut** [comment]
@gsliepen All three nodes using MIPS64el. The operating system is some proprietary one, which seems sharing most code with Linux, but some important syscalls being modified (both calling convention and arguments, for example, one additional argument is added for `execve`, which points at a structure in memory, describing if certain part of the program should be loaded into the DSP or the MIPS). I can build program on it because it has some bug, allowing me to take control, so I think there would not be much people elsewhere to build tinc on it...

And yes, while migrating to `meson` I can feel there do have some advantage points, though there are still some work...

**Comment by eli-schwartz** [comment]
> Well, I may wrong about the autotools part. But ninja and meson are not come together, at least in this project. I have about ~30 nodes running `tinc`, using `1.1` branch. All of those servers have installed `ninja` but no `meson` exists.

Sure, and installing `make` does not mean `autoreconf` exists. I did say that I agree

> Autotools has its own complexity, and the strongest argument people bring for its simplicity is that it's "simple for the end user to use a prepared tarball". Of course it's only simple on non-Windows platforms...

Because a prepared tarball doesn't need autoreconf.

However if you install meson it should be indirectly installing ninja as a dependency...

> Those m4 macros are surely crazy for debugging, but inflation is different from complex.

No, they actually are complex. Just that only the developer, not the end user, interacts with that.

> Even the command is longer than `autoreconf+make`, everyone knows exactly what its doing.

I don't agree.
```
# out of tree build
meson builddir && ninja -C builddir
autoreconf -fi && mkdir builddir && cd builddir && ../configure && make

# in-tree build when supported
meson builddir && ninja -C builddir
autoreconf -fi && ./configure && make
```

...
...

That being said, there's a reason I asked what your specific use cases entail that you're worried about Meson due to. I actually do want to try to help.

> And I know the old build system is not that simple, but at least it is simple enough for most projects. And it run well on all my nodes, including Linux, BSD and some weird linux devices which even have no `python2`.

No python? That's not such a problem: https://mesonbuild.com/FAQ.html#but-i-really-want-a-version-of-meson-that-doesnt-use-python

Take a look at muon, the c99+posix bootstrappable reimplementation of meson, which compiles down to [a 4.5mb statically linked binary](https://muon.build/) that includes `muon samu` so you do not need ninja. This single binary is the entire build system, no need for shell or coreutils or perl/autoconf or make or anything.

It may be significantly easier to install than python. ;)

> Among my nodes, there is 3 special nodes. Those nodes kernel syscall, socket interface and libc are modified so in exact terms they are not Linux anymore. Python and gcc is, of course, not available. Only tmpfs is available, so I have to use "echo -ne \x7fELF" to transfer files. Those took additional amount of work to build and run `tinc`, including patch code for those three different architecture, and patch building scripts to add additional linker arguments. Now I have to dig into those meson snippets and adjust to new build system.

Not completely sure what transferring files via echo here means, are you building tinc locally on that node or cross-compiling and then copying the built binary over?

**Comment by eli-schwartz** [comment]
I needed the following patch to build tinc with muon on linux:

```diff
diff --git a/src/meson.build b/src/meson.build
index d9f7b14b..81b1800f 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -146,7 +146,9 @@ src_tincd = [
   'address_cache.c',
   'autoconnect.c',
   'buffer.c',
-  'compression.h',
+  # this is not needed for dependency ordering, and muon has a bug
+  # that breaks implicit_include_directories if this is here
+  # 'compression.h',
   'conf_net.c',
   'connection.c',
   'control.c',
@@ -208,7 +210,10 @@ if cdata.has('HAVE_SYS_UN_H')
 endif
 
 confdata = configuration_data()
-confdata.merge_from(cdata)
+# muon doesn't yet implement confdata.merge_from
+foreach k: cdata.keys()
+  confdata.set(k, cdata.get(k))
+endforeach
 configure_file(output: 'meson_config.h', configuration: confdata)
 
 have_prefix = '''
@@ -472,4 +477,3 @@ if os_name == 'linux'
 
   benchmark('sptps_speed', exe_sptps_speed, timeout: 90)
 endif
-
```
This is due to two muon issues, one of which I have [a patch for](https://git.sr.ht/~eschwartz/muon/commit/051ff83cd6b1f65403d76b3f0124e76c13c02542).

**Comment by eli-schwartz** [comment]
Update: both issues are fixed in muon upstream. :)
